pwiz_tools/Skyline/Model/AlphaPeptDeep/AlphaPeptDeepLibraryBuilder.cs

```diff
@@ -0,0 +1,405 @@
+                return true;
+            }
+            catch (Exception exception)
+            {
+                progress.UpdateProgress(progressStatus.ChangeErrorException(exception));
+                return false;
+            }
+        }
+
+        private static string CreateDirIfNotExist(string dir)
+        {
+            if (!Directory.Exists(dir))
+            {
+                Directory.CreateDirectory(dir);
+            }
+            return dir;
+        }
+
+        private void RunAlphapeptdeep(IProgressMonitor progress, ref IProgressStatus progressStatus)
+        {
+            progressStatus = progressStatus.ChangeSegments(0, 5);
+
+            PrepareInputFile(progress, ref progressStatus);
+            progressStatus = progressStatus.NextSegment();
+
+            PrepareSettingsFile(progress, ref progressStatus);
+            progressStatus = progressStatus.NextSegment();
+
+            ExecutePeptdeep(progress, ref progressStatus);
+            progressStatus = progressStatus.NextSegment();
+
+            TransformPeptdeepOutput(progress, ref progressStatus);
+            progressStatus = progressStatus.NextSegment();
+
+            ImportSpectralLibrary(progress, ref progressStatus);
+        }
+
+        private void PrepareInputFile(IProgressMonitor progress, ref IProgressStatus progressStatus)
+        {
+            progress.UpdateProgress(progressStatus = progressStatus
+                .ChangeMessage(@"Preparing input file")
+                .ChangePercentComplete(0));
+            
+            var precursorTable = GetPrecursorTable();
+            File.WriteAllLines(InputFilePath, precursorTable);
+
+            progress.UpdateProgress(progressStatus = progressStatus
+                .ChangePercentComplete(100));
+        }
+
+        private IEnumerable<string> GetPrecursorTable()
+        {
+            var result = new List<string>();
+            var header = string.Join(TAB, PrecursorTableColumnNames);
+            result.Add(header);
+
+            // Build precursor table row by row
+            foreach (var peptide in Document.Peptides)
+            {
+                var unmodifiedSequence = peptide.Peptide.Sequence;
+                var modifiedSequence = ModifiedSequence
+                    .GetModifiedSequence(Document.Settings, peptide, IsotopeLabelType.light);
+                var modsBuilder = new StringBuilder();
+                var modSitesBuilder = new StringBuilder();
+
+                for (var i = 0; i < modifiedSequence.ExplicitMods.Count; i++)
+                {
+                    var mod = modifiedSequence.ExplicitMods[i];
+                    if (!mod.UnimodId.HasValue)
+                    {
+                        // TODO(xgwang): update this exception to an Alphapeptdeep specific one
+                        throw new Exception(
+                            @$"Modification {mod} is missing unimod ID, which is required by AlphapeptdeepLibraryBuilder");
```

Comment:
`ModifiedSequence.Modification` does not override ToString(), so you should use `{mod.Name}` here instead of `{mod}`.

It would probably be better to keep track of all of the unsupported modifications that were found, and skip over those peptides, and build a library with as many peptides as possible, and at the end show the list of unsupported modifications which caused some peptides to be skipped.